internal/stacks/stackruntime/internal/stackeval/for_each.go

```diff
@@ -36,22 +37,80 @@ func evaluateForEachExpr(ctx context.Context, expr hcl.Expression, phase EvalPha
 	ty := result.Value.Type()
 
 	const invalidForEachSummary = "Invalid for_each value"
-	const invalidForEachDetail = "The for_each expression must produce either a map of any type or a set of strings. The keys of the map or the set elements will serve as unique identifiers for multiple instances of this embedded stack."
+	const invalidForEachDetail = "The for_each expression must produce either a map of any type or a set of strings. The keys of the map or the set elements will serve as unique identifiers for multiple instances of this resource / data source / provider / component."
```

Comment:
You could modify the function so it accepts a string description the containing block (eg. resource / data source / provider / component). Then the caller can declare itself, and we'll get a more precise error message here.